lib/server.js

```diff
@@ -497,6 +497,8 @@ internals.Server = class {
         finally {
             --this._core.registring;
         }
+
+        return this;
```

Comment:
I believe the idea is that the type of the server returned by `register()` has the effects of the plugin applied.  E.g. if `plugin` supplies decorations, then `await server.register(plugin)` has the type of `server` plus the plugin's decorations.

Reply:
It seems hard to use. How do you write a complex server without getting into a pattern like that:
```ts
const server = Hapi.server();
const server1 = await server.register(plugin1);
const server2 = await server1.register(plugin2);
...
```
As variables cannot really change types once assigned, I see no other way.
The `register` signatures also don't seem to merge with the current server so I'm not sure it's the intention here.

Reply:
Can you provide a usage example for how you actually use this?

Reply:
Understood. If you want to compose plugins, I would suggest merging definitions instead of overwriting them.
I've put up [a simple reproduction](https://www.typescriptlang.org/play?target=99&jsx=0#code/JYOwLgpgTgZghgYwgAgMrQG7QIIAdcA2wCcYwA9iKmKSgN4C+AUE6JLIiulFlAEoQA5sADO7APK4ylEckYs20eEjSZoABQIBXYSAAiEBOSikKIWXSbJryQjtAiA-AC5kAo1AAmAHjFRQggA0yHAgAJ4AfMgAPshaIJ4QMKAQngDcVjYAthBgABbknk6u7sY+fgHBoZExcQlJKelMzExgYbgoBh31IAjAELIAvMgVIIK1o4IA2gC6taVevmD+Y8GTERkK4EqcyJr2IABycDkAatAiZnKZ1iAnEK6TGTbIvJeULiPLAbXxickgVIZFqKDgqfa6dSIADWcEE9ButmhglcENAxzOFzMwK27GUKDRIAAQnARBBvAAVYIAfT0yAgAA9IAlZNxeISDB5TDIopYXlAhKJ2K4ABRknjQVxs6DBchSMwiVwUgCUyEGUQw5GAnlq6ig5CyonJmu1G0RWS0BDIhAgnwARuRyAQIKFfvUAUDEYluolev1isguhAen0Bm7-o1njYBQBHLTAAU5cABvkvGwgQq2x7fMbhhqApppmx5OC4YCfSZ5j2FmwMKuRxGUJD2x3O12xP75z0tNodZAACVLwEJlOCdMZzKKqglUA5hmM3PMaun7O0uk5C+k5iiw0JJLJo8DUQAZMgRYSMRBzlB3iBdWvQFCELD4crNggCKTWWooN5sMvpSgPBCGIRdqFoXlEQFYQxGgQ9xyZYMp0AucuS3EQIhFOxdEVAchxHKkj1leUZE+QCBBgiQSKXDt3UaZVUX1Q0D0Av8TyPHEmCMcwwFsB8QAARlcQcyxHQFeGCOhkByfJClwqSYEdUVVXVZAHSdF07zrBgzW4sQ+IOAAmYT8P47xxJlOQDJw1wpLtOAoGUtUohALQsjtaBkG0s0uJkXjxV4ZdAQAdxXaARTfXyeJGH8BOXOBgrgYB-J-AA6aChXC7DQAEyK9JSmdDPixLkpimcBPSwVYKgLD+MMt8gA) of how this PR behaves when chaining multiple register calls. Notice how server2 doesn't benefit from server1 extensions.

The simple way to fix this is for register not to return `Promise<Server<A> & D>` but `Promise<this & D>`, this way you keep whatever extension was previously applied.